### PR TITLE
[CheckboxGroup] Implement components and hooks

### LINKED ISSUE #213: [CheckboxGroup] Implement CheckboxGroup

### PR DESCRIPTION
**Pull request by atomiks**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/base-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Closes #213

### CODE REVIEW

File: docs/data/base/components/checkbox-group/UnstyledCheckboxGroupIntroduction/css/index.js
Comment (michaldudak) [inline docs/data/base/components/checkbox-group/UnstyledCheckboxGroupIntroduction/css/index.js:1]: Could you please add `'use client'` to all the demos? The new docs will require it.

File: docs/data/base/components/checkbox-group/UnstyledCheckboxGroupIntroduction/css/index.tsx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+const grey = {
+  100: '#E5EAF2',
+  300: '#C7D0DD',
+  500: '#9DA8B7',
+  600: '#6B7A90',
+  800: '#303740',
+  900: '#1C2025',
+};
+
+function useIsDarkMode() {
```
Comment (michaldudak) [inline docs/data/base/components/checkbox-group/UnstyledCheckboxGroupIntroduction/css/index.tsx:56]: Define the dark mode styles by targeting the `body.mode-dark` class, please. It'll make it easier to extract the styles to a separate CSS file in the future.

File: docs/data/base/components/checkbox-group/UnstyledCheckboxGroupNested.tsx
Comment (michaldudak) [inline docs/data/base/components/checkbox-group/UnstyledCheckboxGroupNested.tsx:1]: When I select some items (such as Red and Green) and click on All Colors repeatedly, the selection goes from all to none to Red and Green again. This is quite unexpected (I can't recall seeing such behavior elsewhere). Usually, the parent checkbox toggles between all and none and the partially selected state is lost.
I would make it configurable if you want to keep this behavior.

File: docs/pages/base-ui/api/checkbox-group-root.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
+{
+  "props": {
```
Comment (michaldudak) [inline docs/pages/base-ui/api/checkbox-group-root.json:2]: `allItems` is not documented

File: packages/mui-base/src/CheckboxGroup/Root/CheckboxGroupRootContext.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+export function useCheckboxGroupRootContext() {
+  const context = React.useContext(CheckboxGroupRootContext);
+  if (context === null) {
+    throw new Error('<CheckboxGroup.Label> must be used within <CheckboxGroup.Root>');
+  }
+  return context;
+}
+
+export function useLooseCheckboxGroupRootContext() {
```
Comment (michaldudak) [inline packages/mui-base/src/CheckboxGroup/Root/CheckboxGroupRootContext.ts:16]: Could you use the same pattern as in the Menu (a parameter controlling if the context is required)?

File: packages/mui-base/src/CheckboxGroup/Parent/useCheckboxGroupParent.types.ts
Comment (michaldudak) [inline packages/mui-base/src/CheckboxGroup/Parent/useCheckboxGroupParent.types.ts:1]: Move types to namespaces (per https://github.com/mui/base-ui/issues/517) so we don't introduce more technical debt.

File: docs/data/base/components/checkbox-group/UnstyledCheckboxGroupNested.tsx
Comment (atomiks) [reply docs/data/base/components/checkbox-group/UnstyledCheckboxGroupNested.tsx:1]: This is what the ARIA demo does, but yeah I can make it configurable

Comment (atomiks) [reply docs/data/base/components/checkbox-group/UnstyledCheckboxGroupNested.tsx:1]: Added `preserveChildStates` prop

**Review by colmtuite — APPROVED** [review]

**Pull request merged**